utilities/ggd/lib/dag.go

```diff
@@ -0,0 +1,169 @@
+	pkgs := buildutil.ExpandPatterns(ctx, []string{basePkg + "/..."})
+	g := make(ImportGraph, len(pkgs))
+
+	for pkg := range pkgs {
+		buildPkg, err := ctx.Import(pkg, "", 0)
+		if err != nil {
+			if _, ok := err.(*build.NoGoError); ok {
+				// The package doesn't contain any buildable go files. This can be caused, for
+				// example, when the only go files in a package contain build flags which are
+				// not present in our build context.
+				continue
+			}
+			return nil, fmt.Errorf("could not import package %v: %v", pkg, err)
+		}
+
+		visited := make(map[string]string)
+
+		// ensure we add the pkg to the graph
+		if _, ok := g[pkg]; !ok {
+			g[pkg] = make(ImportSet)
+		}
+
+		for _, path := range buildPkg.Imports {
+			g.maybeAddEdge(ctx, buildPkg, visited, pkg, path)
+		}
+		for _, path := range buildPkg.TestImports {
+			g.maybeAddEdge(ctx, buildPkg, visited, pkg, path)
+		}
+		for _, path := range buildPkg.XTestImports {
+			g.maybeAddEdge(ctx, buildPkg, visited, pkg, path)
+		}
+	}
+
+	return g, nil
+}
+
+func (g ImportGraph) maybeAddEdge(
+	ctx *build.Context, buildPkg *build.Package, visited map[string]string, pkg, path string,
+) {
+	if path == "C" {
+		// Not a real package.
+		return
+	}
+
+	importedPkg, ok := visited[path]
+	if !ok {
+		// It's okay for Import to return an error as not all packages that can be found in
+		// a package will necessarily be present. For example, packages imported only by test
+		// files in vendored packages will not be installed. In the case of an error, Import
+		// always returns a non-nil *Package that contains partial information.
+		importBuildPkg, _ := ctx.Import(path, buildPkg.Dir, build.FindOnly)
+		if importBuildPkg != nil {
+			importedPkg = importBuildPkg.ImportPath
+		} else {
+			importedPkg = path
+		}
+		visited[path] = importedPkg
+		g.addEdge(importedPkg, pkg)
+	}
+}
+
+func (g ImportGraph) addEdge(from, to string) {
+	edges, ok := g[from]
+	if !ok {
+		edges = make(map[string]struct{})
+		g[from] = edges
+	}
+	edges[to] = struct{}{}
+}
+
+// Vertices returns all known vertices in the graph.
+func (g ImportGraph) Vertices() ImportSet {
+	visited := make(ImportSet)
+	for node, impSet := range g {
+		visited[node] = struct{}{}
+		for imp := range impSet {
+			visited[imp] = struct{}{}
+		}
+	}
+	return visited
+}
+
+// Closure return the transitive closure of all packages reachable
+// by starting at the provided paths in the ImportGraph.
+func (g ImportGraph) Closure(paths ...string) (ImportSet, error) {
+	// recursive walk
+	closure := make(ImportSet)
+	for _, p := range paths {
+		if err := g.walk(p, closure); err != nil {
+			return nil, err
+		}
+	}
+
+	// done
+	return closure, nil
+}
+
+func (g ImportGraph) walk(node string, visited ImportSet) error {
+	if _, ok := visited[node]; ok {
+		return nil
+	}
+	if _, ok := g[node]; !ok {
+		// NB(prateek): this happens in the case of un-used deleted packages. Look at testcase7.
+		return DanglingDeleteError{node}
```

Comment:
Why return an error here?

Reply:
need to indicate to main this happened so we can act on it

Reply:
Yea, I guess my question is why do we need to act on it? At least, it doesn't seem like an error to me if I remove a package that wasn't imported.

Reply:
It's not, but I'm being paranoid and rerun the entire test suite if this situation occurs.

Reply:
I see, maybe worth just calling that out?

Reply:
sure

Reply:
done